File: frontend/src/lib/mock/data/metrics.ts
import type { MetricPoint, MetricSeries, ClusterMetrics, NamespaceMetrics } from "@/lib/api/types";
import { mockClusters } from "@/lib/mock/data/clusters";

// ---------------------------------------------------------------------------
// Core generators
// ---------------------------------------------------------------------------

/**
 * Generate a single random metric value between min and max.
 */
export function generateMetricPoint(min: number, max: number): number {
  return min + Math.random() * (max - min);
}

/**
 * Generate a time series of MetricPoint values.
 * Points are spaced 10 seconds apart, ending at the current time.
 */
export function generateTimeSeries(
  points: number,
  min: number,
  max: number,
): MetricPoint[] {
  const now = Date.now();
  const intervalMs = 10_000; // 10 seconds between points
  const series: MetricPoint[] = [];

  // Use a random walk for more realistic-looking data
  let current = min + Math.random() * (max - min);
  const drift = (max - min) * 0.05; // max 5% change per tick

  for (let i = 0; i < points; i++) {
    const delta = (Math.random() - 0.5) * 2 * drift;
    current = Math.max(min, Math.min(max, current + delta));

    series.push({
      timestamp: now - (points - 1 - i) * intervalMs,
      value: Math.round(current * 100) / 100,
    });
  }

  return series;
}

// ---------------------------------------------------------------------------
// Namespace metrics helper
// ---------------------------------------------------------------------------

function generateNamespaceMetrics(
  namespaceName: string,
  memoryUsed: number,
  memoryTotal: number,
  deviceUsed: number,
  deviceTotal: number,
  objects: number,
): NamespaceMetrics {
  return {
    namespace: namespaceName,
    objects,
    memoryUsed,
    memoryTotal,
    deviceUsed,
    deviceTotal,
    readReqs: Math.floor(generateMetricPoint(50_000, 200_000)),
    writeReqs: Math.floor(generateMetricPoint(20_000, 100_000)),
    readSuccess: Math.floor(generateMetricPoint(49_500, 199_500)),
    writeSuccess: Math.floor(generateMetricPoint(19_800, 99_800)),
  };
}

// ---------------------------------------------------------------------------
// Full ClusterMetrics generator
// ---------------------------------------------------------------------------

/**
 * Generate a complete ClusterMetrics snapshot for a connection.
 */
export function generateClusterMetrics(connId: string): ClusterMetrics {
  const cluster = mockClusters[connId];

  // Fallback values if the cluster data is not found
  const nodes = cluster?.nodes ?? [];
  const namespaces = cluster?.namespaces ?? [];

  const totalConnections = nodes.reduce(
    (sum, n) => sum + n.clientConnections,
    0,
  );
  const uptime = nodes.length > 0 ? nodes[0].uptime : 0;

  const timeSeriesPoints = 60; // 10 minutes of data at 10s intervals

  // Build per-namespace metrics and chart series
  const nsMetrics: NamespaceMetrics[] = [];
  const memoryUsageByNs: MetricSeries[] = [];
  const deviceUsageByNs: MetricSeries[] = [];

  const nsColors = ["#0097D3", "#ffe600", "#ff6b35", "#2ecc71", "#9b59b6"];

  for (let i = 0; i < namespaces.length; i++) {
    const ns = namespaces[i];

    nsMetrics.push(
      generateNamespaceMetrics(
        ns.name,
        ns.memoryUsed,
        ns.memoryTotal,
        ns.deviceUsed,
        ns.deviceTotal,
        ns.objects,
      ),
    );

    // Memory usage series: fluctuate around the static memoryUsed value
    const memPctMin = (ns.memoryUsed / ns.memoryTotal) * 100 - 3;
    const memPctMax = (ns.memoryUsed / ns.memoryTotal) * 100 + 3;
    memoryUsageByNs.push({
      name: `memory_${ns.name}`,
      label: `${ns.name} memory`,
      data: generateTimeSeries(timeSeriesPoints, memPctMin, memPctMax),
      color: nsColors[i % nsColors.length],
    });

    // Device usage series
    const devPctMin = (ns.deviceUsed / ns.deviceTotal) * 100 - 2;
    const devPctMax = (ns.deviceUsed / ns.deviceTotal) * 100 + 2;
    deviceUsageByNs.push({
      name: `device_${ns.name}`,
      label: `${ns.name} device`,
      data: generateTimeSeries(timeSeriesPoints, devPctMin, devPctMax),
      color: nsColors[i % nsColors.length],
    });
  }

  // Determine TPS ranges based on connection size
  const isSmallCluster = nodes.length <= 1;
  const readTpsMin = isSmallCluster ? 100 : 300;
  const readTpsMax = isSmallCluster ? 500 : 1500;
  const writeTpsMin = isSmallCluster ? 50 : 150;
  const writeTpsMax = isSmallCluster ? 200 : 800;

  // Aggregate read/write totals from namespace metrics
  const totalReadReqs = nsMetrics.reduce((s, n) => s + n.readReqs, 0);
  const totalWriteReqs = nsMetrics.reduce((s, n) => s + n.writeReqs, 0);
  const totalReadSuccess = nsMetrics.reduce((s, n) => s + n.readSuccess, 0);
  const totalWriteSuccess = nsMetrics.reduce((s, n) => s + n.writeSuccess, 0);

  return {
    connectionId: connId,
    timestamp: Date.now(),
    connected: true,
    uptime,
    clientConnections: totalConnections,
    totalReadReqs,
    totalWriteReqs,
    totalReadSuccess,
    totalWriteSuccess,
    namespaces: nsMetrics,
    readTps: generateTimeSeries(timeSeriesPoints, readTpsMin, readTpsMax),
    writeTps: generateTimeSeries(timeSeriesPoints, writeTpsMin, writeTpsMax),
    connectionHistory: generateTimeSeries(
      timeSeriesPoints,
      isSmallCluster ? 10 : 30,
      isSmallCluster ? 50 : 120,
    ),
    memoryUsageByNs,
    deviceUsageByNs,
  };
}

// ---------------------------------------------------------------------------
// Factory export
// ---------------------------------------------------------------------------

/**
 * Create a metrics generator bound to a specific connection ID.
 * Each call to the returned function produces a fresh snapshot with
 * realistic, slightly randomized values.
 *
 * Usage:
 *   const getMetrics = createMetricsGenerator("conn-1");
 *   const snapshot = getMetrics(); // ClusterMetrics
 */
export function createMetricsGenerator(
  connId: string,
): () => ClusterMetrics {
  return () => generateClusterMetrics(connId);
}

File: backend/src/aerospike_ui_api/mock_data/metrics.py
from __future__ import annotations

import random
import time

from aerospike_ui_api.mock_data.clusters import mock_clusters
from aerospike_ui_api.models.metrics import (
    ClusterMetrics,
    MetricPoint,
    MetricSeries,
    NamespaceMetrics,
)


def _generate_time_series(points: int, min_val: float, max_val: float) -> list[MetricPoint]:
    now = int(time.time() * 1000)
    interval_ms = 10_000
    series: list[MetricPoint] = []

    current = min_val + random.random() * (max_val - min_val)
    drift = (max_val - min_val) * 0.05

    for i in range(points):
        delta = (random.random() - 0.5) * 2 * drift
        current = max(min_val, min(max_val, current + delta))
        series.append(
            MetricPoint(
                timestamp=now - (points - 1 - i) * interval_ms,
                value=round(current * 100) / 100,
            )
        )
    return series


def generate_cluster_metrics(conn_id: str) -> ClusterMetrics:
    cluster = mock_clusters.get(conn_id)
    nodes = cluster.nodes if cluster else []
    namespaces = cluster.namespaces if cluster else []

    total_connections = sum(n.clientConnections for n in nodes)
    uptime = nodes[0].uptime if nodes else 0

    ts_points = 60
    ns_metrics: list[NamespaceMetrics] = []
    memory_usage_by_ns: list[MetricSeries] = []
    device_usage_by_ns: list[MetricSeries] = []
    ns_colors = ["#0097D3", "#ffe600", "#ff6b35", "#2ecc71", "#9b59b6"]

    for i, ns in enumerate(namespaces):
        read_reqs = random.randint(50_000, 200_000)
        write_reqs = random.randint(20_000, 100_000)
        ns_metrics.append(
            NamespaceMetrics(
                namespace=ns.name,
                objects=ns.objects,
                memoryUsed=ns.memoryUsed,
                memoryTotal=ns.memoryTotal,
                deviceUsed=ns.deviceUsed,
                deviceTotal=ns.deviceTotal,
                readReqs=read_reqs,
                writeReqs=write_reqs,
                readSuccess=read_reqs - random.randint(0, 500),
                writeSuccess=write_reqs - random.randint(0, 200),
            )
        )

        mem_pct = (ns.memoryUsed / ns.memoryTotal) * 100
        memory_usage_by_ns.append(
            MetricSeries(
                name=f"memory_{ns.name}",
                label=f"{ns.name} memory",
                data=_generate_time_series(ts_points, mem_pct - 3, mem_pct + 3),
                color=ns_colors[i % len(ns_colors)],
            )
        )

        dev_pct = (ns.deviceUsed / ns.deviceTotal) * 100
        device_usage_by_ns.append(
            MetricSeries(
                name=f"device_{ns.name}",
                label=f"{ns.name} device",
                data=_generate_time_series(ts_points, dev_pct - 2, dev_pct + 2),
                color=ns_colors[i % len(ns_colors)],
            )
        )

    is_small = len(nodes) <= 1
    read_tps = (100, 500) if is_small else (300, 1500)
    write_tps = (50, 200) if is_small else (150, 800)

    return ClusterMetrics(
        connectionId=conn_id,
        timestamp=int(time.time() * 1000),
        connected=True,
        uptime=uptime,
        clientConnections=total_connections,
        totalReadReqs=sum(n.readReqs for n in ns_metrics),
        totalWriteReqs=sum(n.writeReqs for n in ns_metrics),
        totalReadSuccess=sum(n.readSuccess for n in ns_metrics),
        totalWriteSuccess=sum(n.writeSuccess for n in ns_metrics),
        namespaces=ns_metrics,
        readTps=_generate_time_series(ts_points, *read_tps),
        writeTps=_generate_time_series(ts_points, *write_tps),
        connectionHistory=_generate_time_series(
            ts_points,
            10 if is_small else 30,
            50 if is_small else 120,
        ),
        memoryUsageByNs=memory_usage_by_ns,
        deviceUsageByNs=device_usage_by_ns,
    )
